### PR TITLE
[ADD] Module to manage carrier service levels

### PR DESCRIPTION
**Pull request by lfreeke**
This module adds carrier service level.

### CODE REVIEW

**Comment by lfreeke** [comment]
ping @NL66278

**Comment by NL66278** [comment]
@lfreeke Please solve errors. Install pre-commit in your dev env to intercept errors before submitting.

File: delivery_carrier_service_level/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2018 Simone Rubino - Agile Business Group
```
Comment (NL66278) [inline delivery_carrier_service_level/__manifest__.py]: Was this originally made by SImone Rubino? In that case he/she has to be added to the authors. Else this line should be changed.

File: delivery_carrier_service_level/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
```
Comment (NL66278) [inline delivery_carrier_service_level/models/__init__.py]: Make it https:

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2020 Therp BV.
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: Add website to copyright file in the usual way.

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: https:

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+from odoo import fields, models
+
+
+class ServiceLevel(models.Model):
+    _name = "carrier.service.level"
+    _description = "carrier service levels"
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: Description should start with a capital.

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    _order = "carrier_id, code"
+
+    code = fields.Char(string="Carrier Service Level Code")
+    name = fields.Char(string="Carrier Service Level Name", required=True)
+    carrier_id = fields.Many2one("delivery.carrier", string="Carrier")
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: Nitpicking: as this is the main field in the ordering, why not put it first in the list of fields?

File: delivery_carrier_service_level/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+class StockPicking(models.Model):
+    """Extend stock.picking for carrier service level."""
+
+    _name = "stock.picking"
+    _inherit = "stock.picking"
+
+    carrier_service_level_id = fields.Many2one(
+        comodel_name="carrier.service.level",
+        string="Carrier Service Level",
+        store=True,
```
Comment (NL66278) [inline delivery_carrier_service_level/models/stock_picking.py]: store and readonly fields are not needed. This is not a computed field.

**Review by NL66278 — CHANGES_REQUESTED** [review]
:+1: Nice addition but some changes would be welcome and some questions answered.

**Comment by lfreeke** [comment]
>  Nice addition but some changes would be welcome and some questions answered.

@NL66278 Done

File: delivery_carrier_service_level/models/__init__.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
```
Comment (NL66278) [inline delivery_carrier_service_level/models/__init__.py]: lgpl or agpl?

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2013-2020 Therp BV <https://therp.nl>
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: lgpl or agpl?

File: delivery_carrier_service_level/models/__init__.py (hunk 8 above)
Comment (NL66278) [inline delivery_carrier_service_level/models/__init__.py]: lgpl or agpl?

File: delivery_carrier_service_level/models/stock_picking.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2020 Therp BV <https://therp.nl>
```
Comment (NL66278) [inline delivery_carrier_service_level/models/stock_picking.py]: Nitpicking: put a dot at the end of the line, like on the license line. Just more consistent. Not blocking.

File: delivery_carrier_service_level/models/__init__.py (hunk 8 above)
Comment (NL66278) [inline delivery_carrier_service_level/models/__init__.py]: Should still be changed into AGPL.

File: delivery_carrier_service_level/models/carrier_service_level.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+from odoo import fields, models
+
+
+class ServiceLevel(models.Model):
+    _name = "carrier.service.level"
+    _description = "Carrier service levels"
+    _order = "carrier_id, code"
+
+    carrier_id = fields.Many2one("delivery.carrier", string="Carrier")
```
Comment (NL66278) [inline delivery_carrier_service_level/models/carrier_service_level.py]: Make carrier_id and code required as well. Sorry for not remarking this before...

**Review by NL66278 — APPROVED** [review]
:+1: LGTM and working in practice.

File: delivery_carrier_service_level/security/ir.model.access.csv [hunk 12]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_carrier_service_level,carrier.service.level,model_carrier_service_level,stock.group_stock_manager,1,1,1,1
```
Comment (thomaspaulb) [inline delivery_carrier_service_level/security/ir.model.access.csv:2]: Stock managers can set service levels, but shouldn't normal stock users be able to at least read them, lest they get access errors?

Comment (lfreeke) [reply delivery_carrier_service_level/security/ir.model.access.csv:2]: @thomaspaulb I think you are correct. Inventory user is not able to see the configuration but I think the should be able the read the service level on the stock picking

Comment (thomaspaulb) [reply delivery_carrier_service_level/security/ir.model.access.csv:2]: But with your change now, he's also able to write it, eg. change the name of the service level.

Comment (lfreeke) [reply delivery_carrier_service_level/security/ir.model.access.csv:2]: @thomaspaulb Ah you are correct. Fixed it to only read.

**Review by thomaspaulb — APPROVED** [review]
Code review.

**Pull request merged**